pkg/playback/http_playback.go

```diff
@@ -0,0 +1,413 @@
+	}
+
+	switch rr.mode {
+	case Record:
+		rr.httpRecorder.handler(w, r)
+	case Replay:
+		rr.httpReplayer.handler(w, r)
+	case Auto:
+		if rr.isRecordingInAutoMode {
+			rr.httpRecorder.handler(w, r)
+		} else {
+			rr.httpReplayer.handler(w, r)
+		}
+	default:
+		// We should never get here, since all mutations of rr.mode should have already validated the mode value.
+		writeErrorToHTTPResponse(w, fmt.Errorf("got an unexpected playback mode \"%s\". It must be either \"record\", \"replay\", or \"auto\"", rr.mode), 500)
+		return
+	}
+}
+
+// Webhooks are handled slightly differently from outgoing API requests
+func (rr *Server) webhookHandler(w http.ResponseWriter, r *http.Request) {
+	if !rr.cassetteLoaded {
+		w.WriteHeader(400)
+		fmt.Fprint(w, "No cassette is loaded.\n")
+		return
+	}
+
+	switch rr.mode {
+	case Record:
+		rr.httpRecorder.webhookHandler(w, r)
+	case Replay:
+		fmt.Println("Error: webhook endpoint should never be called in replay mode")
+	case Auto:
+		if rr.isRecordingInAutoMode {
+			rr.httpRecorder.webhookHandler(w, r)
+		} else {
+			fmt.Println("Error: webhook endpoint should never be called in replay mode")
+		}
+	}
+}
+
+// note: calling this method always sets cassetteLoaded = false
+func (rr *Server) switchMode(modeString string) error {
+	rr.cassetteLoaded = false
+	switch strings.ToLower(modeString) {
+	case Record:
+		rr.mode = Record
+		return nil
+	case Replay:
+		rr.mode = Replay
+		return nil
+	case Auto:
+		rr.mode = Auto
+		return nil
+	default:
+		return fmt.Errorf("\"%s\" is not a valid playback mode. It must be either \"record\", \"replay\", or \"auto\"", modeString)
+	}
+}
+
+func (rr *Server) setCassetteDir(absoluteCassetteDir string) error {
+	handle, err := os.Stat(absoluteCassetteDir)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return fmt.Errorf("the directory \"%v\" does not exist. Please create it, then try again", absoluteCassetteDir)
+		}
+		return fmt.Errorf("inexpected error when checking cassette directory: %w", err)
+	}
+
+	if !handle.Mode().IsDir() {
+		return fmt.Errorf("the path \"%v\" is not a directory", absoluteCassetteDir)
+	}
+
+	rr.cassetteDirectory = absoluteCassetteDir
+	return nil
+}
+
+// loadCassette() takes a relative path (relative to the rr.cassetteDirectory) to a .yaml file
+// The .yaml file need not exist (unless rr.mode = Replay). If intermediate directories in the relative path do not exist, loadCassette *will* create them.
+func (rr *Server) loadCassette(relativeFilepath string) error {
+	absoluteFilepath := filepath.Join(rr.cassetteDirectory, relativeFilepath)
+
+	var shouldCreateNewFile bool
+
+	switch rr.mode {
+	case Record:
+		shouldCreateNewFile = true
+	case Replay:
+		shouldCreateNewFile = false
+	case Auto:
+		_, err := os.Stat(absoluteFilepath)
+		if os.IsNotExist(err) {
+			shouldCreateNewFile = true
+			rr.isRecordingInAutoMode = true
+		} else {
+			shouldCreateNewFile = false
+			rr.isRecordingInAutoMode = false
+		}
+	}
+
+	if shouldCreateNewFile {
```

Comment:
(Non-blocking suggestion) you can remove a little bit of indirection here:
extract the branches of this if statement into `openCassetteFile`/`openOrCreateCassetteFile` functions. Then instead of setting "shouldCreateNewFile", just call the functions directly.

Then it will be easier to visualize the difference between the cases of your `switch` all in one place.

Reply:
Done! a03d879fc64e5319744cc5974bb4b4e5c40e7ddc